### PR TITLE
Introduce joint rule generator for y flows

### PR DESCRIPTION
**Pull request by timofei-durakov**

### CODE REVIEW

**Review by dpoltavets — APPROVED** [review]

**Review by niksv — APPROVED** [review]

File: src-java/rule-manager/rule-manager-implementation/src/main/java/org/openkilda/rulemanager/factory/generator/flow/MultiTableIngressYRuleGenerator.java [hunk 1]
```diff
@@ -60,13 +62,13 @@
...
-        // TODO(tdurakov): since it's shared meter, this build might be moved outside.
-        SpeakerCommandData meterCommand = buildMeter(flowPath, config, sharedMeterId, sw);
-        if (meterCommand != null) {
-            result.add(meterCommand);
-            command.getDependsOn().add(meterCommand.getUuid());
+        if (generateMeterCommand) {
+            SpeakerCommandData meterCommand = buildMeter(externalMeterCommandUuid, flowPath, config, sharedMeterId, sw);
+            if (meterCommand != null) {
+                result.add(meterCommand);
+                command.getDependsOn().add(externalMeterCommandUuid);
```
Comment (rozdy) [inline src-java/rule-manager/rule-manager-implementation/src/main/java/org/openkilda/rulemanager/factory/generator/flow/MultiTableIngressYRuleGenerator.java:69]: Why `addMeterToInstruction` method call missed here?

File: src-java/rule-manager/rule-manager-implementation/src/main/java/org/openkilda/rulemanager/factory/generator/flow/TransitYRuleGenerator.java [hunk 2]
```diff
@@ -54,13 +56,13 @@
...
-        // TODO(tdurakov): since it's shared meter, this build might be moved outside.
-        SpeakerCommandData meterCommand = buildMeter(flowPath, config, sharedMeterId, sw);
-        if (meterCommand != null) {
-            result.add(meterCommand);
-            command.getDependsOn().add(meterCommand.getUuid());
+        if (generateMeterCommand) {
+            SpeakerCommandData meterCommand = buildMeter(externalMeterCommandUuid, flowPath, config, sharedMeterId, sw);
+            if (meterCommand != null) {
+                result.add(meterCommand);
+                command.getDependsOn().add(externalMeterCommandUuid);
```
Comment (rozdy) [inline src-java/rule-manager/rule-manager-implementation/src/main/java/org/openkilda/rulemanager/factory/generator/flow/TransitYRuleGenerator.java:63]: Why addMeterToInstruction method call missed here too?

File: src-java/rule-manager/rule-manager-implementation/src/main/java/org/openkilda/rulemanager/factory/generator/flow/JointRuleGenerator.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+@Builder
+public class JointRuleGenerator implements RuleGenerator {
+    protected final RuleGenerator firstGenerator;
+    protected final RuleGenerator secondGenerator;
+
+
+    @Override
+    public List<SpeakerCommandData> generateCommands(Switch sw) {
+        List<SpeakerCommandData> commands = firstGenerator.generateCommands(sw);
+        commands.addAll(secondGenerator.generateCommands(sw));
```
Comment (rozdy) [inline src-java/rule-manager/rule-manager-implementation/src/main/java/org/openkilda/rulemanager/factory/generator/flow/JointRuleGenerator.java:35]: Are this two lines of code really needs a separate class? Maybe we can inline them?

File: src-java/rule-manager/rule-manager-implementation/src/main/java/org/openkilda/rulemanager/factory/generator/flow/SingleTableIngressYRuleGenerator.java [hunk 4]
```diff
@@ -60,14 +62,14 @@
...
-            addMeterToInstructions(sharedMeterId, sw, command.getInstructions());
-            result.add(meterCommand);
-            command.getDependsOn().add(meterCommand.getUuid());
+        if (generateMeterCommand) {
+            SpeakerCommandData meterCommand = buildMeter(externalMeterCommandUuid, flowPath, config, sharedMeterId, sw);
+            if (meterCommand != null) {
+                addMeterToInstructions(sharedMeterId, sw, command.getInstructions());
+                result.add(meterCommand);
+                command.getDependsOn().add(externalMeterCommandUuid);
+            }
```
Comment (rozdy) [inline src-java/rule-manager/rule-manager-implementation/src/main/java/org/openkilda/rulemanager/factory/generator/flow/SingleTableIngressYRuleGenerator.java:71]: Looks like the second command in y-flow command pair (with `generateMeterCommand == false`) will not have correct dependency. Actually we need somehow check result of the first generator: are meter command available for the switch before setup dependency for the seccond command. Please add some unit tests on this scenario.

Comment (timofei-durakov) [reply src-java/rule-manager/rule-manager-implementation/src/main/java/org/openkilda/rulemanager/factory/generator/flow/SingleTableIngressYRuleGenerator.java:71]: Nice catch, fixed as we've discussed

**Review by rozdy — APPROVED** [review]

**Pull request merged**